src/NSubstitute/Core/Arguments/ParamsArgumentSpecificationFactory.cs

```diff
@@ -7,57 +7,47 @@ namespace NSubstitute.Core.Arguments
 {
     public class ParamsArgumentSpecificationFactory : IParamsArgumentSpecificationFactory
     {
-        private readonly IDefaultChecker _defaultChecker;
         private readonly IArgumentEqualsSpecificationFactory _argumentEqualsSpecificationFactory;
         private readonly IArrayArgumentSpecificationsFactory _arrayArgumentSpecificationsFactory;
         private readonly IParameterInfosFromParamsArrayFactory _parameterInfosFromParamsArrayFactory;
-        private readonly ISuppliedArgumentSpecificationsFactory _suppliedArgumentSpecificationsFactory;
         private readonly IArrayContentsArgumentSpecificationFactory _arrayContentsArgumentSpecificationFactory;
 
-        public ParamsArgumentSpecificationFactory(IDefaultChecker defaultChecker,
-                                                IArgumentEqualsSpecificationFactory argumentEqualsSpecificationFactory,
+        public ParamsArgumentSpecificationFactory(IArgumentEqualsSpecificationFactory argumentEqualsSpecificationFactory,
                                                 IArrayArgumentSpecificationsFactory arrayArgumentSpecificationsFactory,
                                                 IParameterInfosFromParamsArrayFactory parameterInfosFromParamsArrayFactory,
-                                                ISuppliedArgumentSpecificationsFactory suppliedArgumentSpecificationsFactory,
                                                 IArrayContentsArgumentSpecificationFactory arrayContentsArgumentSpecificationFactory)
         {
-            _defaultChecker = defaultChecker;
             _argumentEqualsSpecificationFactory = argumentEqualsSpecificationFactory;
             _arrayArgumentSpecificationsFactory = arrayArgumentSpecificationsFactory;
             _parameterInfosFromParamsArrayFactory = parameterInfosFromParamsArrayFactory;
-            _suppliedArgumentSpecificationsFactory = suppliedArgumentSpecificationsFactory;
             _arrayContentsArgumentSpecificationFactory = arrayContentsArgumentSpecificationFactory;
         }
 
         public IArgumentSpecification Create(object argument, IParameterInfo parameterInfo, ISuppliedArgumentSpecifications suppliedArgumentSpecifications)
         {
-            if (_defaultChecker.IsDefault(argument, parameterInfo.ParameterType))
```

Comment:
Sorry for the silly question, but why do we no longer need to check is the arg is the default value now?

Reply:
Well, previously we used this check to detect two cases:
1. The argument specification is passed here
2. The `null` value is passed as params array value

I really hated the first point as for me it looks like spreading the responsibilities all over the code (so that a lot of places in code start to know now exactly argument specification placeholder looks). I think that the `ISuppliedArgumentSpecifications` should be the only place to know about that.

Later I found that actually  `suppliedArgumentSpecifications.IsNextFor()` and  `suppliedArgumentSpecifications.AnyFor()` perform the argument value check under the hood, so there is no reason to duplicate that.

That allowed us to keep handling the second scenario only. Given that params array is the reference type, we should check argument for the `null` value only. To make it simpler, I check for the `null` value directly, rather than use the `defaultValueChecker`.

Reply:
Oh so this is just for the case where the entire params arg is specified, rather than matchers supplied _within_ the params? (Sorry, I'm not very familiar with this part of the code)

For example:

```
// Given
Foo SomeMethod(int i, params string[] s);

// Can be called
sub.SomeMethod(2, Arg.Is<string[]>(...)); // match entire params (1)
sub.SomeMethod(2, Arg.Is("hi"), Arg.Is("world")); // match args in the params array (2)
```

This class only handles case `(1)`?

Reply:
Yep. The highlighted if statement (`if (_defaultChecker.IsDefault(argument, parameterInfo.ParameterType))`) was for the case 1 or `sub.SomeMethod(2, null)`. For the case 2 the `else` part of the condition worked.

After the refactoring we still cover both scenarios (see the comments near each block, there I describe the case it handles).